File: packages/pyrbd3/src/pyrbd3/algorithms/sets/_absorb_list.py
from collections import defaultdict
from typing import Iterable, Iterator, Hashable, Set, Dict

class AbsorbList:
    """
    Maintain a family of sets with the following invariant:
      - No set is a subset or superset of another.
      - When inserting a new set:
        * If an existing set is a subset of the new one → ignore the new set.
        * If an existing set is a superset of the new one → remove the supersets and insert the new set.
        * Otherwise, simply add the new set.
    """

    def __init__(self, sets: Iterable[Iterable[Hashable]] | None = None) -> None:
        # Store sets grouped by their size (bucketed by cardinality)
        self.buckets: Dict[int, Set[frozenset]] = defaultdict(set)
        self._count = 0
        if sets:
            self.add_many(sets)

    def __len__(self) -> int:
        return self._count

    def __iter__(self) -> Iterator[set]:
        # Iterate as normal Python sets
        for bucket in self.buckets.values():
            for fs in bucket:
                yield set(fs)

    def __contains__(self, s: Iterable[Hashable]) -> bool:
        fs = frozenset(s)
        return fs in self.buckets[len(fs)]

    def _add_fs(self, fs: frozenset) -> None:
        self.buckets[len(fs)].add(fs)
        self._count += 1

    def _discard_fs(self, fs: frozenset) -> None:
        bucket = self.buckets[len(fs)]
        if fs in bucket:
            bucket.remove(fs)
            self._count -= 1
            if not bucket:
                del self.buckets[len(fs)]

    def add(self, s: Iterable[Hashable]) -> bool:
        """
        Insert a new set `s`.

        Returns:
            True if the family changed (s was added, possibly replacing supersets).
            False if the new set was discarded or already existed.
        """
        fs = frozenset(s)
        m = len(fs)

        # Case 0: exact duplicate
        if fs in self.buckets[m]:
            return False

        # Case 1: if any existing set is a subset of the new one → discard new
        for k in list(self.buckets.keys()):
            if k > m:
                continue
            for exist in self.buckets[k]:
                if exist <= fs:
                    return False

        # Case 2: remove all existing supersets of the new one
        to_remove: list[frozenset] = []
        for k in list(self.buckets.keys()):
            if k < m:
                continue
            for exist in self.buckets[k]:
                if exist >= fs:
                    to_remove.append(exist)

        if to_remove:
            for exist in to_remove:
                self._discard_fs(exist)
            self._add_fs(fs)
            return True
        else:
            # Case 3: independent, just add it
            self._add_fs(fs)
            return True

    def add_many(self, it: Iterable[Iterable[Hashable]]) -> int:
        """
        Batch insert multiple sets.

        Returns:
            Number of times the structure actually changed.
        """
        changed = 0
        # Insert smaller sets first (helps prune supersets early)
        for s in sorted((frozenset(x) for x in it), key=len):
            changed += 1 if self.add(s) else 0
        return changed

    def to_frozensets(self) -> Set[frozenset]:
        """Return all sets as frozensets."""
        out: Set[frozenset] = set()
        for bucket in self.buckets.values():
            out |= set(bucket)
        return out
    
    def to_set_list(self) -> list[set]:
        """Return all sets as list of sets."""
        out: list[set] = []
        for bucket in self.buckets.values():
            for fs in bucket:
                out.append(set(fs))
        return out

    def clear(self) -> None:
        """Remove all stored sets."""
        self.buckets.clear()
        self._count = 0

    def __repr__(self) -> str:
        items = sorted((sorted(fs) for fs in self.to_frozensets()), key=lambda x: (len(x), x))
        return f"AbsorbList({items})"
    

File: packages/pyrbd_core/src/sets/absorb_list.cpp
#include <pyrbd_core/sets.hpp>
#include <algorithm>
#include <unordered_set>

namespace pyrbd_core::sets
{
    // ================================================================
    // AbsorbList — C++ port of Python AbsorbList
    //
    // Maintains a family of sets where no set is a subset or superset
    // of another. Sets are stored sorted for efficient comparison.
    // ================================================================

    // Check if sorted set a ⊆ sorted set b
    bool AbsorbList::isSubsetSorted(const Set& a, const Set& b)
    {
        return std::includes(b.begin(), b.end(), a.begin(), a.end());
    }

    bool AbsorbList::add(const Set& s)
    {
        // Store sets sorted internally
        Set sorted_s = s;
        std::sort(sorted_s.begin(), sorted_s.end());
        size_t m = sorted_s.size();

        // Case 0: exact duplicate
        auto it = buckets_.find(m);
        if (it != buckets_.end())
        {
            for (const auto& existing : it->second)
            {
                if (existing == sorted_s)
                    return false;
            }
        }

        // Case 1: if any existing set (smaller or equal) is a subset of new → discard new
        for (const auto& [k, bucket] : buckets_)
        {
            if (k > m) continue;
            for (const auto& existing : bucket)
            {
                if (isSubsetSorted(existing, sorted_s))
                    return false; // existing ⊆ new → discard new
            }
        }

        // Case 2: remove all existing supersets of the new one
        std::vector<Set> to_remove;
        for (const auto& [k, bucket] : buckets_)
        {
            if (k < m) continue;
            for (const auto& existing : bucket)
            {
                if (isSubsetSorted(sorted_s, existing))
                    to_remove.push_back(existing);
            }
        }

        for (const auto& rem : to_remove)
            discardInternal(rem);

        addInternal(sorted_s);
        return true;
    }

    int AbsorbList::addMany(const std::vector<Set>& sets)
    {
        // Sort by size first (insert smaller sets first to prune more)
        std::vector<Set> sorted_sets = sets;
        std::sort(sorted_sets.begin(), sorted_sets.end(),
                  [](const Set& a, const Set& b) { return a.size() < b.size(); });

        int changed = 0;
        for (const auto& s : sorted_sets)
        {
            if (add(s)) ++changed;
        }
        return changed;
    }

    std::vector<Set> AbsorbList::toSetList() const
    {
        std::vector<Set> result;
        result.reserve(count_);
        for (const auto& [_, bucket] : buckets_)
        {
            for (const auto& s : bucket)
                result.push_back(s);
        }
        return result;
    }

    void AbsorbList::clear()
    {
        buckets_.clear();
        count_ = 0;
    }

    void AbsorbList::addInternal(const Set& s)
    {
        buckets_[s.size()].push_back(s);
        ++count_;
    }

    void AbsorbList::discardInternal(const Set& s)
    {
        auto it = buckets_.find(s.size());
        if (it == buckets_.end()) return;

        auto& bucket = it->second;
        auto pos = std::find(bucket.begin(), bucket.end(), s);
        if (pos != bucket.end())
        {
            bucket.erase(pos);
            --count_;
            if (bucket.empty())
                buckets_.erase(it);
        }
    }

} // namespace pyrbd_core::sets
